File: iosApp/iosApp/core/logging/tests/LoggableTest.swift
import Testing
import shared

struct LoggableTest {
    class TestClass: Loggable {
        let logger: any Logging = shared.CoreUtilsModule.shared.recordingLogger()
        var recLogger: RecordingLogging? { logger as? RecordingLogging }
    }

    private let sut = TestClass()

    @Test func testConditionTest() async throws {
        #expect(sut.recLogger != nil)
    }

    @Test func tagTest() async throws {
        #expect("TestClass" == sut.tag)
    }

    @Test func logDTest() async throws {
        let msg = "Log Debug"
        sut.logD(message: msg)

        #expect(sut.tag == sut.recLogger?.latestRecordLog?.tag)
        #expect(msg == sut.recLogger?.latestRecordLog?.message)
        #expect("d" == sut.recLogger?.latestRecordLog?.level)
    }

    @Test func logITest() async throws {
        let msg = "Log Info"
        sut.logI(message: msg)

        #expect(sut.tag == sut.recLogger?.latestRecordLog?.tag)
        #expect(msg == sut.recLogger?.latestRecordLog?.message)
        #expect("i" == sut.recLogger?.latestRecordLog?.level)
    }

    @Test func logWTest() async throws {
        let msg = "Log Warning"
        sut.logW(message: msg)

        #expect(sut.tag == sut.recLogger?.latestRecordLog?.tag)
        #expect(msg == sut.recLogger?.latestRecordLog?.message)
        #expect("w" == sut.recLogger?.latestRecordLog?.level)
    }

    @Test func logETest() async throws {
        let msg = "Log Error"
        sut.logE(message: msg)

        #expect(sut.tag == sut.recLogger?.latestRecordLog?.tag)
        #expect(msg == sut.recLogger?.latestRecordLog?.message)
        #expect("e" == sut.recLogger?.latestRecordLog?.level)
    }
}

File: core/logging/implementation/src/commonTest/kotlin/de/stefan/lang/utils/logging/implementation/logging/LoggableTest.kt
package de.stefan.lang.utils.logging.implementation.logging

import de.stefan.lang.utils.logging.contract.Loggable
import de.stefan.lang.utils.logging.contract.RecordLog
import de.stefan.lang.utils.logging.contract.RecordingLogging
import de.stefan.lang.utils.logging.implementation.recording.RecordingLogger
import de.stefan.lang.utils.logging.implementation.silent.SilentLogger
import kotlin.test.Test
import kotlin.test.assertEquals

class LoggableTest {
    class TestLoggable : Loggable {
        override val logger: RecordingLogging = RecordingLogger(SilentLogger())
        val currentRecordLog: RecordLog? get() = logger.latestRecordLog
    }

    private val expectedTag = "TestLoggable"
    private val message = "Test Message"

    @Test
    fun `correct Tag should be returned`() {
        val sut = TestLoggable()
        assertEquals("TestLoggable", sut.tag)
    }

    @Test
    fun `logD should return correct values`() {
        val sut = TestLoggable()
        sut.logD(message)

        assertEquals(RecordLog(expectedTag, "d", message), sut.currentRecordLog)
    }

    @Test
    fun `logI should return correct values`() {
        val sut = TestLoggable()
        sut.logI(message)

        assertEquals(RecordLog(expectedTag, "i", message), sut.currentRecordLog)
    }

    @Test
    fun `logW should return correct values`() {
        val sut = TestLoggable()
        sut.logW(message)

        assertEquals(RecordLog(expectedTag, "w", message), sut.currentRecordLog)
    }

    @Test
    fun `logE should return correct values`() {
        val sut = TestLoggable()
        sut.logE(message)

        assertEquals(RecordLog(expectedTag, "e", message), sut.currentRecordLog)
    }
}
